src/types/redis_bloom_chain.cc

```diff
@@ -0,0 +1,184 @@
+
+  *ret = 0;
+  if (found && mode == ReadWriteMode::READ) {
+    *ret = 1;
+  } else if (!found && mode == ReadWriteMode::WRITE) {
+    auto batch = storage_->GetWriteBatchBase();
+    *ret = 1;
+    block_split_bloom_filter.InsertHash(h);
+    batch->Put(bf_key, block_split_bloom_filter.GetData());
+    return storage_->Write(storage_->DefaultWriteOptions(), batch->GetWriteBatch());
+  }
+
+  return rocksdb::Status::OK();
+}
+
+rocksdb::Status BloomChain::Reserve(const Slice &user_key, uint32_t capacity, double error_rate, uint16_t expansion) {
+  std::string ns_key = AppendNamespacePrefix(user_key);
+
+  LockGuard guard(storage_->GetLockManager(), ns_key);
+  BloomChainMetadata bloom_chain_metadata;
+  rocksdb::Status s = getBloomChainMetadata(ns_key, &bloom_chain_metadata);
+  if (!s.ok() && !s.IsNotFound()) return s;
+  if (!s.IsNotFound()) {
+    return rocksdb::Status::InvalidArgument("the key already exists");
+  }
+
+  return createBloomChain(ns_key, error_rate, capacity, expansion, &bloom_chain_metadata);
+}
+
+rocksdb::Status BloomChain::Add(const Slice &user_key, const Slice &item, int *ret) {
+  std::string ns_key = AppendNamespacePrefix(user_key);
+  LockGuard guard(storage_->GetLockManager(), ns_key);
+
+  BloomChainMetadata metadata;
+  rocksdb::Status s = getBloomChainMetadata(ns_key, &metadata);
+
+  if (s.IsNotFound()) {
+    rocksdb::Status create_s =
+        createBloomChain(ns_key, kBFDefaultErrorRate, kBFDefaultInitCapacity, kBFDefaultExpansion, &metadata);
+    if (!create_s.ok()) return create_s;
+  } else if (!s.ok()) {
+    return s;
+  }
+
+  std::vector<std::string> bf_key_list;
+  getBFKeyList(ns_key, metadata, &bf_key_list);
+
+  auto batch = storage_->GetWriteBatchBase();
+  WriteBatchLogData log_data(kRedisBloomFilter, {"insert"});
+  batch->PutLogData(log_data.Encode());
+
+  bool item_exist = false;
+  std::string item_string = item.ToString();
+
+  // check
+  for (int i = metadata.n_filters - 1; i >= 0; --i) {  // TODO: to test which direction for searching is better
+    s = bloomCheckAdd(bf_key_list[i], item_string, ReadWriteMode::READ, ret);
+    if (*ret == 1) {
+      item_exist = true;
+      break;
+    }
+  }
+
+  // insert
+  if (!item_exist) {
+    if (metadata.size + 1 > metadata.GetCapacity()) {  // TODO: scaling would be supported later
+      return rocksdb::Status::Aborted("filter is full");
+    }
+    s = bloomCheckAdd(bf_key_list.back(), item_string, ReadWriteMode::WRITE, ret);
+    metadata.size += *ret;
+  }
+
+  std::string sb_chain_metadata_bytes;
+  metadata.Encode(&sb_chain_metadata_bytes);
+  batch->Put(metadata_cf_handle_, ns_key, sb_chain_metadata_bytes);
+
+  return storage_->Write(storage_->DefaultWriteOptions(), batch->GetWriteBatch());
+}
+
+rocksdb::Status BloomChain::Exist(const Slice &user_key, const Slice &item, int *ret) {
+  std::string ns_key = AppendNamespacePrefix(user_key);
+  LockGuard guard(storage_->GetLockManager(), ns_key);
```

Comment:
I think we can split the check and add logic. For the existing operation, it's unnecessary to lock for adding the missing bloom filter, because the BloomFilter read operations would be much larger than the write operations in most of scenarios. So it'd be better to non-lock for that.

Reply:
Actually I think we can delay locking when checking the "last" bloom-filter.

During insert:
1. Acquire n_filters
2. (without-lock)  check filter `1...n_filters-1`
3. (lock) check final filter

This is possible because `1...n_filters-1` is immutable. This can be regard as a future optimization.